fili-core/src/main/java/com/yahoo/bard/webservice/servlet/RoleImpersonationFilter.java

```diff
@@ -0,0 +1,220 @@
+
+        // if okta, populate `isUserInRole`
+        String fromId = requestContext.getHeaders().getFirst("From");
+        if (fromId == null || fromId.equals("")) {
+            return;
+        }
+
+        if (!isAuthorizedToImpersonate(requestContext.getSecurityContext())) {
+            throwUnauthorizedImpersonationError(requestContext, fromId);
+        }
+
+        Matcher matcher = EMAIL_PATTERN.matcher(fromId);
+        boolean matches = matcher.find();
+        if (!matches) {
+            throwMalformedFromHeaderError(fromId);
+        }
+        String userId = matcher.group(1);
+        String domain = matcher.group(2);
+
+        List<String> roles = null;
+        try {
+            roles = getRolesForId(userId);
+            roles = roles.stream().map(s -> s.toLowerCase(Locale.US)).collect(Collectors.toList());
+        } catch (IOException e) {
+            String error = String.format("Failed authorization with from header: (%s)", fromId);
+            LOG.error(error, e);
+            throwImpersonatedUserAuthorizationError(fromId);
+        }
+        List<String> finalRoles = roles;
+
+        SecurityContext originalSecurityContext = requestContext.getSecurityContext();
+        Principal principal = originalSecurityContext.getUserPrincipal();
+        DelegatedPrincipal delegatedPrincipal = new DelegatedPrincipal() {
+            @Override
+            public Principal deletegatedFrom() {
+                return principal;
+            }
+
+            @Override
+            public String getName() {
+                return userId;
+            }
+        };
+        requestContext.setSecurityContext(new SecurityContext() {
+            @Override
+            public Principal getUserPrincipal() {
+                return delegatedPrincipal;
+            }
+
+            @Override
+            public boolean isUserInRole(String role) {
+                String conformedRole = aliases.getOrDefault(role, role).toLowerCase(Locale.US);
+                return finalRoles.contains(conformedRole);
+            }
+
+            @Override
+            public boolean isSecure() {
+                return originalSecurityContext.isSecure();
+            }
+
+            @Override
+            public String getAuthenticationScheme() {
+                return originalSecurityContext.getAuthenticationScheme();
+            }
+        });
+    }
+
+    /**
+     * An exception occurred where the user requesting authorization does not have the privilege to impersonate.
+     *
+     * @param requestContext  The request context for the current request.
+     * @param fromId  The attempted impersonation.
+     *
+     * @throws IOException with an WebApplication FORBIDDEN status
+     */
+    private void throwUnauthorizedImpersonationError(
+            final ContainerRequestContext requestContext,
+            final String fromId
+    ) {
+        Response.ResponseBuilder builder = null;
```

Comment:
Why the early declaration? Why not just declare this on line 189 when we build it?